schematic/schemas/data_model_parser.py

```diff
@@ -0,0 +1,319 @@
+        model_dict = self.gather_csv_attributes_relationships(model_df)
+
+        return model_dict
+
+class DataModelJSONLDParser():
+    def __init__(
+        self,
+        ):
+        # Instantiate DataModelRelationships
+        self.dmr = DataModelRelationships()
+        # Load relationships dictionary.
+        self.rel_dict = self.dmr.define_data_model_relationships()
+
+    def gather_jsonld_attributes_relationships(
+        self,
+        model_jsonld: List[dict]) -> Dict:
+        '''
+        Args:
+            model_jsonld: list of dictionaries, each dictionary is an entry in the jsonld data model
+        Returns:
+            attr_rel_dictionary: dict,
+                {Node Display Name: 
+                    {Relationships: {
+                                     CSV Header: Value}}}
+        Notes:
+            - Unlike a CSV the JSONLD might already have a base schema attached to it.
+              So the attributes:relationship dictionary for importing a CSV vs JSONLD may not match.
+            - It is also just about impossible to extract attributes explicitly. Using a dictionary should avoid duplications.
+            - This is a promiscuous capture and will create an attribute for each model entry.
+            - Currently only designed to capture the same information that would be encoded in CSV, 
+                can be updated in the future.
+        TODO: 
+            - Find a way to delete non-attribute keys, is there a way to reliable distinguish after the fact?
+            - Right now, here we are stripping contexts, will need to track them in the future.
+        '''
+        
+        # Retrieve relevant JSONLD keys.
+        jsonld_keys_to_extract = ['label', 'subClassOf', 'id', 'displayName']
+        label_jsonld_key, subclassof_jsonld_key, id_jsonld_key, dn_jsonld_key = [self.rel_dict[key]['jsonld_key']
+                                                    for key in jsonld_keys_to_extract ]
+        
+        # Gather all labels from the model.
+        model_labels = [v[label_jsonld_key] for v in model_jsonld]
+
+        attr_rel_dictionary = {}
+        # Move through each entry in the jsonld model
+        for entry in model_jsonld:
+            # Get the label of the entry
+            try:
+                # Get the entry display name (if recorded)
+                entry_name = entry[dn_jsonld_key]
+            except:
+                # If no display name, get the label.
+                entry_name = entry[label_jsonld_key]
+
+            # If the entry is an attribute that has not already been added to the dictionary, add it.
+            if entry_name not in attr_rel_dictionary.keys():
+                attr_rel_dictionary.update({entry_name: {'Relationships': {}}})
+            
+            # Add relationships for each attribute
+            # 
+            # Go through each defined relationship type (key) and its attributes (val)
+            for key, val in self.rel_dict.items():
```

Comment:
Can we break this for block up? It's very complicated.

Reply:
Extracted parsing to its own function so it reads a bit nicer.